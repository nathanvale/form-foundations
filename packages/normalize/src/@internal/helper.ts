export function isLastChar(value, char) {
  return !!value && value.substr(value.length - 1) === char
}

export function shouldDeleteLastChar(value, previousValue, normalizedChar) {
  if (previousValue) {
    const isBackspacing = value.length < previousValue.length
    if (isBackspacing) {
      if (isLastChar(previousValue, normalizedChar)) {
        return previousValue.substr(0, previousValue.length - 1)
      }
    }
  }
}

export function noChangeOnDeleteLastChar(value, previousValue) {
  if (previousValue) {
    const isBackspacing = value.length < previousValue.length
    if (isBackspacing) {
      return value
    }
  }
}
