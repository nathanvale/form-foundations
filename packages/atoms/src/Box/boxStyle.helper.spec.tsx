import * as boxStyleHelper from './boxStyle.helper'

describe('Box Helpers', () => {
  describe('marginTop', () => {
    it('should output style for marginTop at a XS breakpoint', () => {
      const expected = boxStyleHelper.marginTop()(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginTop at a SM breakpoint', () => {
      const expected = boxStyleHelper.marginTop('SM')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginTop at a MD breakpoint', () => {
      const expected = boxStyleHelper.marginTop('MD')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginTop at a LG breakpoint', () => {
      const expected = boxStyleHelper.marginTop('LG')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginTop at a XL breakpoint', () => {
      const expected = boxStyleHelper.marginTop('XL')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should throw an error with an invalid breakpoint', () => {
      function test() {
        boxStyleHelper.marginTop('invalid breakpoint')
      }
      expect(test).toThrowError('Breakpoints must be SM, MD, LG or XL')
    })
  })
  describe('marginRight', () => {
    it('should output style for marginRight at a XS breakpoint', () => {
      const expected = boxStyleHelper.marginRight()(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginRight at a SM breakpoint', () => {
      const expected = boxStyleHelper.marginRight('SM')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginRight at a MD breakpoint', () => {
      const expected = boxStyleHelper.marginRight('MD')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginRight at a LG breakpoint', () => {
      const expected = boxStyleHelper.marginRight('LG')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginRight at a XL breakpoint', () => {
      const expected = boxStyleHelper.marginRight('XL')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should throw an error with an invalid breakpoint', () => {
      function test() {
        boxStyleHelper.marginRight('invalid breakpoint')
      }
      expect(test).toThrowError('Breakpoints must be SM, MD, LG or XL')
    })
  })
  describe('marginBottom', () => {
    it('should output style for marginBottom at a XS breakpoint', () => {
      const expected = boxStyleHelper.marginBottom()(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginBottom at a SM breakpoint', () => {
      const expected = boxStyleHelper.marginBottom('SM')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginBottom at a MD breakpoint', () => {
      const expected = boxStyleHelper.marginBottom('MD')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginBottom at a LG breakpoint', () => {
      const expected = boxStyleHelper.marginBottom('LG')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginBottom at a XL breakpoint', () => {
      const expected = boxStyleHelper.marginBottom('XL')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should throw an error with an invalid breakpoint', () => {
      function test() {
        boxStyleHelper.marginBottom('invalid breakpoint')
      }
      expect(test).toThrowError('Breakpoints must be SM, MD, LG or XL')
    })
  })
  describe('marginLeft', () => {
    it('should output style for marginLeft at a XS breakpoint', () => {
      const expected = boxStyleHelper.marginLeft()(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginLeft at a SM breakpoint', () => {
      const expected = boxStyleHelper.marginLeft('SM')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginLeft at a MD breakpoint', () => {
      const expected = boxStyleHelper.marginLeft('MD')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginLeft at a LG breakpoint', () => {
      const expected = boxStyleHelper.marginLeft('LG')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for marginLeft at a XL breakpoint', () => {
      const expected = boxStyleHelper.marginLeft('XL')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should throw an error with an invalid breakpoint', () => {
      function test() {
        boxStyleHelper.marginLeft('invalid breakpoint')
      }
      expect(test).toThrowError('Breakpoints must be SM, MD, LG or XL')
    })
  })
  describe('paddingX', () => {
    it('should output style for paddingX at a XS breakpoint', () => {
      const expected = boxStyleHelper.paddingX()(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for paddingX at a SM breakpoint', () => {
      const expected = boxStyleHelper.paddingX('SM')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for paddingX at a MD breakpoint', () => {
      const expected = boxStyleHelper.paddingX('MD')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for paddingX at a LG breakpoint', () => {
      const expected = boxStyleHelper.paddingX('LG')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for paddingX at a XL breakpoint', () => {
      const expected = boxStyleHelper.paddingX('XL')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should throw an error with an invalid breakpoint', () => {
      function test() {
        boxStyleHelper.paddingX('invalid breakpoint')
      }
      expect(test).toThrowError('Breakpoints must be SM, MD, LG or XL')
    })
  })
  describe('paddingY', () => {
    it('should output style for paddingY at a XS breakpoint', () => {
      const expected = boxStyleHelper.paddingY()(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for paddingY at a SM breakpoint', () => {
      const expected = boxStyleHelper.paddingY('SM')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for paddingY at a MD breakpoint', () => {
      const expected = boxStyleHelper.paddingY('MD')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for paddingY at a LG breakpoint', () => {
      const expected = boxStyleHelper.paddingY('LG')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should output style for paddingY at a XL breakpoint', () => {
      const expected = boxStyleHelper.paddingY('XL')(100)
      expect(expected).toMatchSnapshot()
    })
    it('should throw an error with an invalid breakpoint', () => {
      function test() {
        boxStyleHelper.paddingY('invalid breakpoint')
      }
      expect(test).toThrowError('Breakpoints must be SM, MD, LG or XL')
    })
  })
  describe('size', () => {
    it('should return empty when property is empty', () => {
      expect(boxStyleHelper.dangerousStyleValue('height', null)).toEqual('')
      expect(boxStyleHelper.dangerousStyleValue('height', '')).toEqual('')
      expect(boxStyleHelper.dangerousStyleValue('height', true)).toEqual('')
      expect(boxStyleHelper.dangerousStyleValue('height', false)).toEqual('')
    })

    it('should add "px" suffix to attributes', () => {
      expect(boxStyleHelper.dangerousStyleValue('height', 10)).toEqual('10px')
      expect(boxStyleHelper.dangerousStyleValue('width', 20)).toEqual('20px')
    })

    it('should not add "px" suffix to the unitless properties', () => {
      expect(boxStyleHelper.dangerousStyleValue('flex', 1)).toEqual('1')
      expect(boxStyleHelper.dangerousStyleValue('zIndex', 99)).toEqual('99')
    })

    it('should keep the same value if it is custom property', () => {
      expect(
        boxStyleHelper.dangerousStyleValue('custom-property', '100%', true),
      ).toEqual('100%')
    })
  })
})
