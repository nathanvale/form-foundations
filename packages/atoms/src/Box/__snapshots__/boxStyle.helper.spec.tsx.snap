// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box Helpers marginBottom should output style for marginBottom at a LG breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "992px",
  ") {
      ",
  "
        ",
  "
      margin-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginBottom should output style for marginBottom at a MD breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "768px",
  ") {
      ",
  "
        ",
  "
      margin-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginBottom should output style for marginBottom at a SM breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "576px",
  ") {
      ",
  "
        ",
  "
      margin-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginBottom should output style for marginBottom at a XL breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "1200px",
  ") {
      ",
  "
        ",
  "
      margin-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginBottom should output style for marginBottom at a XS breakpoint 1`] = `
Array [
  "
      margin-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
]
`;

exports[`Box Helpers marginLeft should output style for marginLeft at a LG breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "992px",
  ") {
      ",
  "
        ",
  "
      margin-left: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginLeft should output style for marginLeft at a MD breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "768px",
  ") {
      ",
  "
        ",
  "
      margin-left: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginLeft should output style for marginLeft at a SM breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "576px",
  ") {
      ",
  "
        ",
  "
      margin-left: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginLeft should output style for marginLeft at a XL breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "1200px",
  ") {
      ",
  "
        ",
  "
      margin-left: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginLeft should output style for marginLeft at a XS breakpoint 1`] = `
Array [
  "
      margin-left: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
]
`;

exports[`Box Helpers marginRight should output style for marginRight at a LG breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "992px",
  ") {
      ",
  "
        ",
  "
      margin-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginRight should output style for marginRight at a MD breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "768px",
  ") {
      ",
  "
        ",
  "
      margin-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginRight should output style for marginRight at a SM breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "576px",
  ") {
      ",
  "
        ",
  "
      margin-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginRight should output style for marginRight at a XL breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "1200px",
  ") {
      ",
  "
        ",
  "
      margin-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginRight should output style for marginRight at a XS breakpoint 1`] = `
Array [
  "
      margin-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
]
`;

exports[`Box Helpers marginTop should output style for marginTop at a LG breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "992px",
  ") {
      ",
  "
        ",
  "
      margin-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginTop should output style for marginTop at a MD breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "768px",
  ") {
      ",
  "
        ",
  "
      margin-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginTop should output style for marginTop at a SM breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "576px",
  ") {
      ",
  "
        ",
  "
      margin-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginTop should output style for marginTop at a XL breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "1200px",
  ") {
      ",
  "
        ",
  "
      margin-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers marginTop should output style for marginTop at a XS breakpoint 1`] = `
Array [
  "
      margin-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
]
`;

exports[`Box Helpers paddingX should output style for paddingX at a LG breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "992px",
  ") {
      ",
  "
        ",
  "
      padding-left: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers paddingX should output style for paddingX at a MD breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "768px",
  ") {
      ",
  "
        ",
  "
      padding-left: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers paddingX should output style for paddingX at a SM breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "576px",
  ") {
      ",
  "
        ",
  "
      padding-left: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers paddingX should output style for paddingX at a XL breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "1200px",
  ") {
      ",
  "
        ",
  "
      padding-left: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers paddingX should output style for paddingX at a XS breakpoint 1`] = `
Array [
  "
      padding-left: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-right: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
]
`;

exports[`Box Helpers paddingY should output style for paddingY at a LG breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "992px",
  ") {
      ",
  "
        ",
  "
      padding-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers paddingY should output style for paddingY at a MD breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "768px",
  ") {
      ",
  "
        ",
  "
      padding-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers paddingY should output style for paddingY at a SM breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "576px",
  ") {
      ",
  "
        ",
  "
      padding-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers paddingY should output style for paddingY at a XL breakpoint 1`] = `
Array [
  "
    @media only screen and (min-width: ",
  "1200px",
  ") {
      ",
  "
        ",
  "
      padding-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
  ";
    }
  ",
]
`;

exports[`Box Helpers paddingY should output style for paddingY at a XS breakpoint 1`] = `
Array [
  "
      padding-bottom: calc(",
  "4px",
  " * ",
  "100",
  ");
      padding-top: calc(",
  "4px",
  " * ",
  "100",
  ");
    ",
]
`;
